### PR TITLE
Added UnmarshalOpt to allow a best effort unmarshal

### PR DESCRIPTION
**Pull request by lgomez9**
Added an option to allow a failing unmarshal will now pass - instead of failing, it will simply log the errors and continue the unmarshaling process. This is so that we can generate diffs for the entirety of the SetRequest, even if there's a compliance error.

### CODE REVIEW

File: ytypes/unmarshal.go [hunk 1]
```diff
@@ -29,6 +29,14 @@ type UnmarshalOpt interface {
 	IsUnmarshalOpt()
 }
 
+// BestEffortUnmarshal is an unmarshal option that 'ignores' errors while unmarshalling,
```
Comment (wenovus) [inline ytypes/unmarshal.go:32]: ignores -> accumulates?

File: ytypes/unmarshal.go [hunk 2]
```diff
@@ -29,6 +29,14 @@ type UnmarshalOpt interface {
...
+// and continues the unmarshaling process. A failing unmarshal will still fail, but it will
+// fail as a series of error reports.
```
Comment (wenovus) [inline ytypes/unmarshal.go:34]: What kind of error type will be returned?

One simple way of doing the error report is gnmi's errlist package: https://pkg.go.dev/github.com/openconfig/gnmi/errlist

However for more detailed error reporting you might consider what is being used in ygnmi:
https://github.com/openconfig/ygnmi/blob/613186339658c020a54b31ebe8e751e94e10ab43/ygnmi/unmarshal.go#L65-L74

In fact ygnmi does a very similar operation of unmarshalling Notifications to a GoStruct, and I might be able to leverage your effort here if you choose this route. I have been planning to use UnmarshalNotifications in ygnmi to do this.

File: ytypes/unmarshal.go [hunk 3]
```diff
@@ -29,6 +29,14 @@ type UnmarshalOpt interface {
...
+type BestEffortUnmarshal struct{}
```
Comment (wenovus) [inline ytypes/unmarshal.go:35]: Perhaps make this unexported for now and export it after the functionality is ready in case we want to create another release before then?

File: ytypes/unmarshal.go [hunk 4]
```diff
@@ -29,6 +29,40 @@ type UnmarshalOpt interface {
 	IsUnmarshalOpt()
 }
 
+// ComplianceErrors contains the compliance errors encountered from an Unmarshal operation.
+type ComplianceErrors struct {
+	// Generic errors for now, until we make a decision on what specific types of errors should
+	// be returned
+	Errors []error
```
Comment (wenovus) [inline ytypes/unmarshal.go:36]: https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences
Comments should begin with the name of the thing being described and end in a period:

File: ytypes/gnmi.go [hunk 5]
```diff
@@ -62,9 +63,10 @@ func UnmarshalNotifications(schema *Schema, ns []*gpb.Notification, opts ...Unma
 //
 // If an error occurs during unmarshalling, schema.Root may already be
 // modified. A rollback is not performed.
-func UnmarshalSetRequest(schema *Schema, req *gpb.SetRequest, opts ...UnmarshalOpt) error {
+func UnmarshalSetRequest(schema *Schema, req *gpb.SetRequest, opts ...UnmarshalOpt) (*ComplianceErrors, error) {
```
Comment (wenovus) [inline ytypes/gnmi.go:66]: Let's add some tests for this flag.

File: ytypes/gnmi.go [hunk 6]
```diff
@@ -87,10 +89,10 @@ func UnmarshalSetRequest(schema *Schema, req *gpb.SetRequest, opts ...UnmarshalO
 	if err := deletePaths(schema.SchemaTree[nodeName], node, prefix, req.Delete, preferShadowPath); err != nil {
```
Comment (wenovus) [inline ytypes/gnmi.go:89]: Should this apply to deletes as well?

File: ytypes/gnmi.go [hunk 7]
```diff
@@ -35,7 +36,7 @@ import (
 //
 // If an error occurs during unmarshalling, schema.Root may already be
 // modified. A rollback is not performed.
-func UnmarshalNotifications(schema *Schema, ns []*gpb.Notification, opts ...UnmarshalOpt) error {
+func UnmarshalNotifications(schema *Schema, ns []*gpb.Notification, opts ...UnmarshalOpt) (*ComplianceErrors, error) {
```
Comment (wenovus) [inline ytypes/gnmi.go:39]: So this would be a backwards-incompatible change. I think it's better to make ComplianceErrors satisfy the error interface and return it instead. So essentially the flag controls whether this structure is returned.

**Review by wenovus — CHANGES_REQUESTED** [review]

File: ytypes/unmarshal.go (hunk 2 above)
Comment (wenovus) [reply ytypes/unmarshal.go]: For this we've decided to return an error type that's to-be-filled.

File: ytypes/unmarshal.go (hunk 3 above)
Comment (wenovus) [reply ytypes/unmarshal.go]: To enable early usage we've decided to keep this exported and just have the MVP functionality of accumulating errors in an error report.

File: ytypes/unmarshal.go [hunk 8]
```diff
@@ -29,6 +29,49 @@ type UnmarshalOpt interface {
...
+		return &ComplianceErrors{Errors: errs}
+	}
+
+	c.Errors = append(c.Errors, errs...)
+	return c
+}
+
+// BestEffortUnmarshal is an unmarshal option that accumulates errors while unmarshalling,
+// and continues the unmarshaling process. An unmarshal will no longer fail, but
+// will return an error report.
```
[the remark this answers is not in the document]
Comment (lgomez9) [reply ytypes/unmarshal.go:69]: Ack.

File: ytypes/unmarshal.go [hunk 9]
```diff
@@ -29,6 +30,49 @@ type UnmarshalOpt interface {
...
+	}
+
+	c.Errors = append(c.Errors, errs...)
+	return c
+}
+
+// BestEffortUnmarshal is an unmarshal option that accumulates errors while unmarshalling,
+// and continues the unmarshaling process. An unmarshal now return a ComplianceErrors struct,
+// instead of a single error.
+type bestEffortUnmarshal struct{}
```
Comment (wenovus) [inline ytypes/unmarshal.go:71]: Since we have decided on an MVP, if you intend to use this prior to adding more functionality then feel free to make this exported.

File: ytypes/gnmi_test.go [hunk 10]
```diff
@@ -496,6 +496,48 @@ func TestUnmarshalSetRequest(t *testing.T) {
 			}},
 		},
 		wantErr: true,
+	}, {
+		desc: "mix of an error and a non-error update with best-effort flag",
```
Comment (wenovus) [inline ytypes/gnmi_test.go:500]: For testing I recommend using a complete set-up but testing **two** errors in each of 1) delete 2) replace 3) update and then checking that all six errors are present. This way you get coverage of all six modifications in the handling logic.

**Review by wenovus — CHANGES_REQUESTED** [review]
PTAL at the presubmit failures.

File: ytypes/unmarshal.go (hunk 9 above)
Comment (lgomez9) [reply ytypes/unmarshal.go:71]: Ack.

File: ytypes/gnmi_test.go (hunk 10 above)
Comment (lgomez9) [reply ytypes/gnmi_test.go:500]: I think I understand - I'll upload a new version in a few, let me know if I've misunderstood.

File: ytypes/gnmi_test.go [hunk 11]
```diff
@@ -504,6 +641,16 @@ func TestUnmarshalSetRequest(t *testing.T) {
 			if gotErr := err != nil; gotErr != tt.wantErr {
 				t.Fatalf("got error: %v, want: %v", err, tt.wantErr)
 			}
+			if gotErr := err != nil; gotErr && hasBestEffortUnmarshal(tt.inUnmarshalOpts) {
```
Comment (wenovus) [inline ytypes/gnmi_test.go:644]: nit: can just `if err != nil && hasBestEffort`

File: ytypes/gnmi_test.go [hunk 12]
```diff
@@ -496,6 +498,141 @@ func TestUnmarshalSetRequest(t *testing.T) {
 			}},
 		},
 		wantErr: true,
+	}, {
+		desc: "mix of an error and a non-error update with best-effort flag",
```
Comment (wenovus) [inline ytypes/gnmi_test.go:502]: Actually I was thinking of putting all the errors into a single SetRequest. If you just add another test case with everything in these three cases or just merge all of these into one then this LGTM.

File: ytypes/gnmi_test.go (hunk 11 above)
Comment (lgomez9) [reply ytypes/gnmi_test.go:644]: Makes way more sense.

File: ytypes/gnmi_test.go (hunk 12 above)
Comment (lgomez9) [reply ytypes/gnmi_test.go:502]: Moving them all into one.

**Review by wenovus — APPROVED** [review]
Thanks!

File: ytypes/unmarshal.go [hunk 13]
```diff
@@ -29,6 +30,49 @@ type UnmarshalOpt interface {
...
+		for _, e := range c.Errors {
+			b.WriteString("\n\t")
+			b.WriteString(e.Error())
+		}
+	} else {
+		b.WriteString(" None")
+	}
+	b.WriteString("\n")
+	return b.String()
+}
```
Comment (wenovus) [inline ytypes/unmarshal.go:57]: Can you make it so that we have coverage for this? A good way is to use this: https://pkg.go.dev/github.com/openconfig/gnmi/errdiff#Substring and just check for the presence of "Noncompliance errors" whenever `hasBestEffortUnmarshal` returns true. I think it makes sense to just add this option for one other random passing test to get coverage for the nil case. Thanks!

**Review by wenovus — CHANGES_REQUESTED** [review]
Missed reviewing the coverage, can you add coverage for one part of the code and the rest looks good.

File: ytypes/unmarshal.go (hunk 13 above)
Comment (lgomez9) [reply ytypes/unmarshal.go:57]: Is it okay to have the coverage in gnmi_test.go? Currently, the BestEffortUnmarshal only has an effect when used on the UnmarshalSetRequest function. I could also add functionality for this to the unmarshalGeneric function if you think that'd be a good idea.

Comment (wenovus) [reply ytypes/unmarshal.go:57]: It's ok that `ytypes.Unmarshal` doesn't support this option. Actually it would be good to add logic there to error out if this option is provided saying it's not supported. It's somewhat involved to get it working there since it deals with JSON and there are a lot of places where errors can happen so I'm ok with not supporting it.

As for coverage of the `String()` method I'm fine with either having direct coverage in unmarshal_test.go or via Substring.

Oh yes, since this also affects UnmarshalNotifications can you add a test case for that as well? Thanks.

Comment (lgomez9) [reply ytypes/unmarshal.go:57]: Added!

**Review by wenovus — APPROVED** [review]
Thank you, LGTM

**Comment by lgomez9** [comment]
Sorry, thought I had run gofmt, but forgot to.

**Review by wenovus — APPROVED** [review]
Thanks! merged
